### PR TITLE
Fix race-condition in _lf_replace_template_token

### PR DESCRIPTION
**Pull request by erlingrj**
This addresses the problem described in #2081 

The problem was that we moved tokens from the trigger structs to the action structs outside a critical section. This created a race condition with external asynchronus contexts trying to schedule events onto the very same action struct.

This simple solution does the moving of tokens inside i critical section. I think another solution might be to re-think the design. Particularily. 
1. Why does `lf_schedule` put the token on BOTH the event queue AND the action struct? The runtime will later move the token from event queue -> trigger struct and then in the reaction preambles, it moves it from the trigger struct to the action struct.
2. Why is the token moved from trigger-struct to action-struct in the code-generated preamble. It would be more natural to do that from wihin the runtime (which is in a critical section)

### CODE REVIEW

**Comment by erlingrj** [comment]
@edwardalee

**Comment by erlingrj** [comment]
I had to include `platform.h` for the reaction bodies in order to code-generate a critical section in the reaction preamble. It suggests that maybe another solution is better. Like having the runtime move the token instead of code-generating it.

**Review by lhstrh — APPROVED** [review]
You found a needle in a haystack. Terrific job, @erlingrj.

File: core/src/main/java/org/lflang/generator/c/CReactionGenerator.java [hunk 1]
```diff
@@ -561,11 +563,13 @@ private static String generateActionVariablesInReaction(
                 + " != NULL && "
                 + tokenPointer
                 + "->value != NULL);",
+            "lf_critical_section_enter(self->base.environment);",
```
Comment (edwardalee) [inline core/src/main/java/org/lflang/generator/c/CReactionGenerator.java:566]: I think the critical section needs to include the two operations above, setting is_present and has_value.  Also the setting of the value below.

**Comment by erlingrj** [comment]
@lhstrh we can wait a little with merging. There might be a better way of addressing this

File: core/src/main/java/org/lflang/generator/c/CReactionGenerator.java [hunk 2]
```diff
@@ -343,11 +343,13 @@ public static String generateForwardBody(
                 + ")self->_lf__"
                 + actionName
                 + ".tmplt.token->value;",
+            "lf_critical_section_enter(self->base.environment);",
```
Comment (edwardalee) [inline core/src/main/java/org/lflang/generator/c/CReactionGenerator.java:346]: I think the critical section needs to include the operations above, setting the value.  Also the setting of is_present below.

**Review by edwardalee — CHANGES_REQUESTED** [review]
Good catch, but I think the critical section needs to be a little bigger or the action struct could become inconsistent.

**Comment by erlingrj** [comment]
@edwardalee I expanded the critical sections according to your comments.

**Comment by erlingrj** [comment]
@edwardalee, have you seen this test failure of `ChainWithDelay` before?
```
 Federate 2: FATAL ERROR: Attempted to insert reactions for a trigger that had an intended tag that was in the past. This should not happen under centralized coordination. Intended tag: (2500000, 0). Current tag: (3000000, 0)).
```

**Comment by erlingrj** [comment]
I was able to reproduce it here and I have the trace of federated communication. Should we discuss it in a separate issue? It might be related to fixes you are working on? The RTI erroneously gives a PTAG to a federate

**Comment by edwardalee** [comment]
Yes, I'm seeing these kinds of failures. There is an easy solution, but it will mask a more fundamental problem.  The easy solution is to prohibit the use of PTAG except for zero-delay cycles.  We will want to do that eventually, but first we should fix the handling of PTAG because with this easy fix, it will be hard to produce these errors.

**Comment by erlingrj** [comment]
Separate issue for the ChainWithDelay PTAG problem: https://github.com/lf-lang/lingua-franca/issues/2084

**Pull request merged**